c7n/resources/directconnect.py

```diff
@@ -18,3 +22,66 @@ class resource_type(TypeInfo):
         universal_taggable = object()
 
     augment = universal_augment
+
+
+@resources.register('directconnect-vif')
+class DirectConnectVirtualInterface(QueryResourceManager):
+
+    class resource_type(TypeInfo):
+        service = 'directconnect'
+        enum_spec = ('describe_virtual_interfaces', 'virtualInterfaces', None)
+        id = filter_name = 'virtualInterfaceId'
+        name = 'virtualInterfaceName'
+        filter_type = 'scalar'
+        arn_type = "dxvif"
+        universal_taggable = object()
+
+
+@resources.register('directconnect-gateway')
+class DirectConnectGateway(QueryResourceManager):
+
+    class resource_type(TypeInfo):
+        service = 'directconnect'
+        enum_spec = ('describe_direct_connect_gateways', 'directConnectGateways', None)
+        id = filter_name = 'directConnectGatewayId'
+        name = 'directConnectGatewayName'
+        filter_type = 'scalar'
+        arn_type = "dx-gateway"
+
+
+@DirectConnectGateway.filter_registry.register('associations')
+class DirectConnectGatewayAssociations(ValueFilter):
+    """Filters Direct Connect Gateways by attached Associations
+
+    This pulls a Direct Connect Gateway resource's Assocations and adds
+    them as a new field to each resource, and allows us to filter resources
+    based on these Associations.
+
+    An example of this is filtering Gateway resources to find any instances
+    associated with resources in an AWS account outside of a provided whitelist.
+    """
+    schema = type_schema('associations', rinherit=ValueFilter.schema)
+    permissions = ('directconnect:DescribeDirectConnectGatewayAssociations',)
+
+    def process(self, resources, event=None):
+
+        client = local_session(self.manager.session_factory).client('directconnect')
+
+        def _augment(r):
+            try:
+                r['directConnectGatewayAssociations'] = self.manager.retry(
```

Comment:
we generally do augments like this in a c7n: prefix annotation to avoid ambiguity with resource builtin attributes.

that would then need a match function override, such that the associations filter's  key is done directly against the annotation.

do you have an example of the type of attribute check your trying to do against the association?

Reply:
Got it.

I want to check if a Gateway has any associations located in AWS accounts (i.e. `directConnectGatewayAssociations` list of resources, each with an `ownerAccount` field) that are/aren't included in our provided list of whitelisted account IDs. The unit test for this filter (see https://github.com/cloud-custodian/cloud-custodian/pull/6860/files#diff-35e38c8bbb0a856be46e375096c83388841d1f4f08915d931fa72991a4e552faR58 ) has a policy somewhat similar to what I am trying to support.